### PR TITLE
Updated PythonHL generation to remove exception while inheriting device having Enum attribute.

### PR DESCRIPTION
**Pull request by mesnehal**
Updated the PythonHL code generation to support:
1. Enum attributes from inherited device
2. Enum attribute initialization having label inconsistent to Python standards

Attached a zip having sample BaseDevice and InheritedDevice.
[ResolvedIneritanceIssue.zip](https://github.com/tango-controls/pogo/files/4349344/ResolvedIneritanceIssue.zip)

### CODE REVIEW

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend [hunk 1]
```diff
@@ -292,7 +292,7 @@ class PythonUtils {
     def String checkEnumLabels (Attribute attr){
     	var flag = "valid"
     	for(label: attr.enumLabels){
-    	 	if(label.contains("-")||label.contains("!")||label.contains("#")||label.contains("@")||label.contains("%")||label.contains("$")||label.charAt(0).toString.matches("[0-9]")){
+    	 	if(label.contains("-")||label.contains("+")||label.contains("=")||label.contains("%")||label.contains("*")||label.charAt(0).toString.matches("[0-9]")){
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend]: Why the change to the characters?  Agree that `+=*` are problems, but `!#@$` seem problematic too.   Does some other part of Pogo prevent them getting here?

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonTypeDefinitions.java [hunk 2]
```diff
@@ -319,7 +319,13 @@ public static String defaultValueHL (Attribute attr) {
 		if (attr.getDataType() instanceof EncodedType)			def_val =  "['', '']";
 		if (attr.getDataType() instanceof EnumType)
 		{
-			def_val = toFirstUpper(attr.getName())+ "."+ attr.getEnumLabels().get(0);
+			if(attr.getEnumLabels().get(0).contains("-") || attr.getEnumLabels().get(0).contains("%") || attr.getEnumLabels().get(0).contains("*")
+					|| attr.getEnumLabels().get(0).contains("+") || attr.getEnumLabels().get(0).contains("=")) {
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonTypeDefinitions.java]: This could be a lot shorter, and easier to read if the first label was put in a local variable before all the checks.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend [hunk 3]
```diff
@@ -250,6 +251,7 @@ def enumClasses(PogoDeviceClass cls) '''
 		"""Python enumerated type for «attr.name.toFirstUpper» attribute."""
 		«ENDIF»
 	«ENDIF»
+	«ENDIF»
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend]: The inherited class has too many blank lines after the attributes section header:
```python
    # ----------
    # Attributes
    # ----------




    testMode = attribute(
        dtype=TestMode,
    )
```

Just a single line, like the base device, would be better:
```python
    # ----------
    # Attributes
    # ----------

    healthState = attribute(
        dtype=HealthState,
    )
```

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend [hunk 4]
```diff
@@ -1362,6 +1363,7 @@ class Test«cls.name»(object):
             assert tango_context.device.«attr.name» == «attr.defaultValueHL»
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend]: This results in something like:
```python
        assert tango_context.device.healthState == HealthState.OK
```
which is good, but the `HealthState` class is not imported in the test class, so it would fail.
Adding `from BaseDevice import HealthState` after the other imports would help.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend (hunk 1 above)
Comment (mesnehal) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend]: Yes the characters !#@$ are prevented from the user interface. POGO UI does not allow to enter these characters as part of Enum labels. Whereas for +=* are not prevented from UI so handled them as part of code.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonTypeDefinitions.java (hunk 2 above)
Comment (mesnehal) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonTypeDefinitions.java]: Agree. Will do that change.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend (hunk 3 above)
Comment (mesnehal) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend]: Yes agree. Will update that.

File: org.tango.pogo.pogo_gui/src/org/tango/pogo/pogo_gui/DeviceClass.java [hunk 5]
```diff
@@ -429,6 +429,7 @@ public String getSourceFile(String lang) {
             case "cpp":
                 return path + pogoClass.getName() + ".cpp";
             case "python":
+            	return path + pogoClass.getName() + ".py";
```
[the remark this answers is not in the document]
Comment (mesnehal) [reply org.tango.pogo.pogo_gui/src/org/tango/pogo/pogo_gui/DeviceClass.java]: Yes agree. Few of my changes were giving the same exception while generating the code. But somehow generation is not giving that exception now after certain new changes. I will revert this change back to earlier one. Thanks.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend (hunk 4 above)
Comment (ajoubertza) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend]: I still don't see something like `from BaseDevice import HealthState` in the latest changes.

Comment (mesnehal) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend]: I have added import statement in Device_test.py file for importing BaseDevice. Importing other Enum classes from the base device is a bigger activity and could be handled as a separate issue.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend [hunk 6]
```diff
@@ -1281,6 +1281,7 @@ from PyTango import DevState
 «IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»
         «protectedAreaHL(cls, "test_additional_imports")»
 «ENDIF»
+«cls.inheritedAdditionalImportHL»
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend]: I don't understand why this is being added.  If I understand correctly, this will result in something like `from BaseDevice import BaseDevice`.   In the generated tests, the `BaseDevice` class isn't used, so it is unnecessary.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend (hunk 4 above)
Comment (ajoubertza) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend]: OK, the import of the Enum classes for tests can be handled later.  It was already "broken" before this PR.

Comment (mesnehal) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonHlProjectUtils.xtend]: Okay Anton. Thanks.

**Comment by Pascal-Verdier** [comment]
Will be available in next Pogo release

**Comment by ajoubertza** [comment]
Hi @Pascal-Verdier.
I see you have made a 9.6.30 release tag.  Will you be uploading the new build to https://bintray.com/tango-controls/maven/Pogo soon?  Please do.  Thanks.

**Comment by Pascal-Verdier** [comment]
Yes. I tried.
But build a jar file failed:
**The following artifacts could not be resolved: org.eclipse:help:jar:3.3.0-v20070524**

Something has changed ?

**Comment by mesnehal** [comment]
Hi Pascal, locally, I am able to generate the jar file. Changes done are reflected in this PR.
Is there something, I can check? Please do let me know.

**Comment by Pascal-Verdier** [comment]
I cannot understand.
The pom.xml files did not changed except release number but it fails with message:
**Failed to execute goal on project Pogo: Could not resolve dependencies for project org.tango.tools.pogo.gui:Pogo:jar:9.6.31-SNAPSHOT: The following artifacts could not be resolved: org.eclipse:help:jar:3.3.0-v20070524, org.eclipse:jface:jar:3.3.0-I20070606-0010, org.eclipse:swt:jar:3.3.0-v3346, org.eclipse.core:expressions:jar:3.3.0-v20070606-0010, org.eclipse.jface:databinding:jar:1.1.0-I20070606-0010, org.eclipse.core:databinding:jar:1.0.0-I20070606-0010: Could not find artifact org.eclipse:help:jar:3.3.0-v20070524 in JBoss (https://repository.jboss.org/nexus/content/groups/public-jboss)**

May be it works on your computer because the dependency is still in your maven cache ?

**Comment by mesnehal** [comment]
Hi, maybe computer's cache have the dependencies but I am not sure. I have pulled the updated(merged) pogo master code and am able to generate the jar. 
Just to let you know, I have done the new setup on my Windows machine before raising this PR (installing Eclipse IDE and building the maven, etc). 

Is there something specific I can do to help resolve this? Please do let me know.

**Comment by Pascal-Verdier** [comment]
Hi
After adding lines in my ~/.m2/settings.xml file, I have been able to generate a new jar file.
[Pogo-9.6.31.jar](https://bintray.com/tango-controls/maven/download_file?file_path=org%2Ftango%2Ftools%2Fpogo%2Fgui%2FPogo%2F9.6.31%2FPogo-9.6.31.jarl)
Regards

**Comment by mesnehal** [comment]
That's great!!
Thanks for the update :)
